python/runtime/dbapi/connection.py

```diff
@@ -66,19 +70,38 @@ def success(self):
 
     @abstractmethod
     def close(self):
-        """Close the ResultSet explicitly, release any resource incurred by this query"""
+        """Close the ResultSet explicitly, release any resource incurred by this query
+        implementation should support close multi-times"""
         pass
 
 
-class Connection(ABC):
+@six.add_metaclass(ABCMeta)
+class Connection(object):
     """Base class for DB connection
 
     Args:
         conn_uri: a connection uri in the schema://name:passwd@host/path?params format
 
     """
     def __init__(self, conn_uri):
-        self.conn_uri = conn_uri
+        self.uristr = conn_uri
+        self.uripts = self._parse_uri()
+        self.params = parse_qs(
+            self.uripts.query,
+            keep_blank_values=True,
+        )
+        self.params["database"] = self.uripts.path.strip("/")
+        for k, l in self.params.items():
+            if len(l) == 1:
+                self.params[k] = self.params[k][0]
```

Comment:
Not quite sure. Is it safe in Python if we change `self.params` when iterating?

Reply:
I think it's ok to modify the value, but we should not del or add keys. I tried some case, it just works.